hydromt/data_adapter/data_adapter_base.py

```diff
@@ -1,16 +1,51 @@
 """BaseModel for DataAdapter."""
 
-from typing import Any, Dict, Union
+from datetime import timedelta
+from typing import Any, Dict, List, Optional
 
-from pydantic import BaseModel, Field
+from pydantic import BaseModel, ConfigDict, Field
+
+from hydromt._typing import TimeRange
 
 
 class DataAdapterBase(BaseModel):
     """BaseModel for DataAdapter."""
 
+    model_config = ConfigDict(extra="forbid")
+
     unit_add: Dict[str, Any] = Field(default_factory=dict)
     unit_mult: Dict[str, Any] = Field(default_factory=dict)
     rename: Dict[str, str] = Field(default_factory=dict)
-    attrs: Dict[str, Any] = Field(default_factory=dict)
-    meta: Dict[str, Any] = Field(default_factory=dict)
-    nodata: Union[float, int, Dict[str, Union[float, int]], None] = Field(default=None)
+
+    def to_source_timerange(
+        self,
+        time_range: TimeRange,
+    ) -> TimeRange:
+        """
+        Tranform a DataSource timerange to the source-native timerange.
+
+        args:
+            time_range: TimeRange
+                start and end datetime.
+        """
+        if dt := self.unit_add.get("time"):
+            # subtract from source unit add
+            return (time - timedelta(seconds=dt) for time in time_range)
```

Comment:
would we need to manage timezones here? (genuine question, I have no idea)

Reply:
This is about the time in the dataset, so I guess that really depends on the data. We can always use this unit_add for simple procedures, otherwise we would need to do some preprocessing. Definitely out of scope for this PR though!

Reply:
agreed.